src/Gui/MainWindow.cpp

```diff
@@ -150,11 +150,92 @@ class CustomMessageEvent : public QEvent
     int _timeout;
 };
 
+class DimensionWidget : public QPushButton, WindowParameter
+{
+    Q_OBJECT
+
+public:
+    explicit DimensionWidget(QWidget* parent): QPushButton(parent), WindowParameter("Units")
+    {
+        setFlat(true);
+        setText(qApp->translate("Gui::MainWindow", "Dimension"));
+        setMinimumWidth(120);
+
+        //create the action buttons
+        auto* menu = new QMenu(this);
+        auto* actionGrp = new QActionGroup(menu);
+        int num = static_cast<int>(Base::UnitSystem::NumUnitSystemTypes);
+        for (int i = 0; i < num; i++) {
+            QAction* action = menu->addAction(QStringLiteral("UnitSchema%1").arg(i));
+            actionGrp->addAction(action);
+            action->setCheckable(true);
+            action->setData(i);
+        }
+        QObject::connect(actionGrp, &QActionGroup::triggered, this, [this](QAction* action) {
+            int userSchema = action->data().toInt();
+            // Set and save the Unit System
+            Base::UnitsApi::setSchema(static_cast<Base::UnitSystem>(userSchema));
+            getWindowParameter()->SetInt("UserSchema", userSchema);
+            // Update the application to show the unit change
+            Gui::Application::Instance->onUpdate();
+        } );
+        setMenu(menu);
+        retranslateUi();
+        unitChanged();
+        getWindowParameter()->Attach(this);
+    }
+
+    ~DimensionWidget()
+    {
+        getWindowParameter()->Detach(this);
+    }
+
+    void OnChange(Base::Subject<const char*> &rCaller, const char * sReason) override
+    {
+        Q_UNUSED(rCaller)
+        if (strcmp(sReason, "UserSchema") == 0) {
+            unitChanged();
+        }
+    }
+
+    void changeEvent(QEvent *event) override
+    {
+        if (event->type() == QEvent::LanguageChange) {
+            retranslateUi();
+        }
+        else {
+            QPushButton::changeEvent(event);
+        }
+    }
+
+private:
+    void unitChanged(void)
+    {
+        int userSchema = getWindowParameter()->GetInt("UserSchema", 0);
+        auto actions = menu()->actions();
+        if(Q_UNLIKELY(userSchema < 0 || userSchema >= actions.size())) {
+            userSchema = 0;
+        }
+        actions[userSchema]->setChecked(true);
+    }
+
+    void retranslateUi() {
+        auto actions = menu()->actions();
+        int maxSchema = static_cast<int>(Base::UnitSystem::NumUnitSystemTypes);
+        assert(actions.size() <= maxSchema);
+        for(int i = 0; i < maxSchema ; i++)
+        {
+            actions[i]->setText(qApp->translate("Gui::Dialog::DlgSettingsUnits",
+                    Base::UnitsApi::getDescription(static_cast<Base::UnitSystem>(i))));
+        }
+    }
+};
+
 // -------------------------------------
 // Pimpl class
 struct MainWindowP
 {
-    QToolButton* sizeLabel;
+    QPushButton* sizeLabel;
```

Comment:
Why not DimensionWidget here?

Reply:
Roughly, to help keep encapsulation under control. :)
The `MainWindow` class doesn't need to know more than it is using a QPushButton, then the rest is intended to stay internal stuff.
It has also the advantage that changing the `DimensionWidget` internals is "transparent" for the `MainWindow`.
However your interrogation seems legit and I'm not 100% sure to be right on this. @wwmayer may you please advise the right way ?

Reply:
As said we can leave it for now. Once we decide to move the class(es) we can do the same as Qt and create private headers.

Reply:
@wwmayer I'm afraid you didn't get the point (maybe my mistake). :)
Simply the question is : is that OK for you to declare `sizeLabel` as `QPushButton` while it will actually be initialized as a `DimensionWidget` -- that inherits `QPushButton` -- taking as argument that `MainWindowP` doesn't need to know more than that ? Thx

Reply:
> I'm afraid you didn't get the point (maybe my mistake). :)

Sorry, my bad. Using QPushButton here is totally fine.